File: services/core-api/canton/prepare_execute.go
package canton

// PrepareExecuteParams mirrors the Canton dApp SDK prepareExecute parameter schema
// (CIP-103 prepareExecuteAndWait).
type PrepareExecuteParams struct {
	CommandID          string        `json:"commandId"`
	Commands           []interface{} `json:"commands"`
	ActAs              []string      `json:"actAs"`
	ReadAs             []string      `json:"readAs"`
	DisclosedContracts []interface{} `json:"disclosedContracts"`
}

// ToPrepareExecuteParams maps a PreparedLedgerCommand into WalletConnect-ready params.
func ToPrepareExecuteParams(prepared PreparedLedgerCommand) PrepareExecuteParams {
	loopSubmit := prepared.LoopSubmit
	commands := loopSubmit.Commands
	if commands == nil {
		commands = []interface{}{}
	}
	disclosed := loopSubmit.DisclosedContracts
	if disclosed == nil {
		disclosed = []interface{}{}
	}

	commandID := loopSubmit.CommandID
	if commandID == "" {
		commandID = prepared.CommandID
	}

	return PrepareExecuteParams{
		CommandID:          commandID,
		Commands:           commands,
		ActAs:              loopSubmit.ActAs,
		ReadAs:             loopSubmit.ReadAs,
		DisclosedContracts: disclosed,
	}
}

File: apps/web/lib/canton/prepare-execute.ts
import type { PrepareExecuteParams } from '@canton-network/dapp-sdk';
import type {
  LoopWalletSubmitPayload,
  PrepareExecutePayload,
  PreparedLedgerCommand,
} from '../../types/ledger';

function normalizeCommands(
  commands: LoopWalletSubmitPayload['commands'],
): PrepareExecuteParams['commands'] {
  if (Array.isArray(commands)) {
    return commands as PrepareExecuteParams['commands'];
  }

  if (commands && typeof commands === 'object') {
    const entries = Object.entries(commands as Record<string, unknown>).sort(
      ([left], [right]) => Number(left) - Number(right),
    );
    return entries.map(([, command]) => command) as PrepareExecuteParams['commands'];
  }

  return [];
}

export function toPrepareExecuteParams(
  prepared: PreparedLedgerCommand,
): PrepareExecuteParams {
  const loopSubmit = prepared.loop_submit;

  return {
    commandId: loopSubmit.commandId ?? prepared.command_id,
    commands: normalizeCommands(loopSubmit.commands),
    actAs: loopSubmit.actAs,
    readAs: loopSubmit.readAs,
    disclosedContracts: (loopSubmit.disclosedContracts ??
      []) as PrepareExecuteParams['disclosedContracts'],
  };
}

/** Prefer API `prepare_execute` when present (transfer-nft endpoint). */
export function resolvePrepareExecuteParams(
  prepared: PreparedLedgerCommand,
  prepareExecute?: PrepareExecutePayload,
): PrepareExecuteParams {
  if (prepareExecute) {
    return {
      commandId: prepareExecute.commandId,
      commands: prepareExecute.commands as PrepareExecuteParams['commands'],
      actAs: prepareExecute.actAs,
      readAs: prepareExecute.readAs,
      disclosedContracts: (prepareExecute.disclosedContracts ??
        []) as PrepareExecuteParams['disclosedContracts'],
    };
  }

  return toPrepareExecuteParams(prepared);
}
